playbooks/openstack/inventory.py

```diff
@@ -16,52 +16,74 @@
 import shade
 
 
-def base_openshift_inventory(cluster_hosts):
+def base_openshift_inventory(cluster_hosts, inventory,
+                             docker_storage_mountpoints):
     '''Set the base openshift inventory.'''
-    inventory = {}
+    nodes = set()
+    osev3 = set()
+
+    inventory['cluster_hosts'] = {'hosts': []}
+    inventory['masters'] = {'hosts': []}
+    inventory['etcd'] = {'hosts': []}
+    inventory['infra_hosts'] = {'hosts': []}
+    inventory['app'] = {'hosts': []}
+    inventory['glusterfs'] = {'hosts': []}
+    inventory['dns'] = {'hosts': []}
+    inventory['lb'] = {'hosts': []}
+    inventory['localhost'] = {'ansible_connection': 'local'}
+
+    for server in cluster_hosts:
+        if 'group' in server.metadata:
+            group = server.metadata.get('group')
+            if group not in inventory:
+                inventory[group] = {'hosts': []}
+            inventory[group]['hosts'].append(server.name)
 
-    masters = [server.name for server in cluster_hosts
-               if server.metadata['host-type'] == 'master']
+        inventory['cluster_hosts']['hosts'].append(server.name)
 
-    etcd = [server.name for server in cluster_hosts
-            if server.metadata['host-type'] == 'etcd']
-    if not etcd:
-        etcd = masters
+        if server.metadata['host-type'] == 'master':
+            inventory['masters']['hosts'].append(server.name)
+            nodes.add(server.name)
+            osev3.add(server.name)
 
-    infra_hosts = [server.name for server in cluster_hosts
-                   if server.metadata['host-type'] == 'node' and
-                   server.metadata['sub-host-type'] == 'infra']
+        if server.metadata['host-type'] == 'etcd':
+            inventory['etcd']['hosts'].append(server.name)
+            osev3.add(server.name)
```

Comment:
Does this mean that etcd servers are **not** part of the nodes set? It looks like their hosts will be added to inventory a few lines down, but will they/should they be in the node set as well?